compiler/rustc_typeck/src/check/method/mod.rs

```diff
@@ -511,6 +516,31 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         };
         obligations.extend(traits::predicates_for_generics(cause.clone(), self.param_env, bounds));
```

Comment:
It might be possible to stash the type from the expectation in the `BinOp` cause-code then edit the code that suggests "add `std::ops::Mul`" to add the associated type there, in case inference is broken by the current change (or if the perf hit is too massive). I'll try to think about if either is a really valid worry, though.

Reply:
>  in case inference is broken by the current change 

I guess not necessarily, since we're only applying this when we have `Expectation::ExpectHasType`, not the other ones which are there to guide coercion... (I think)